### ISSUE TITLE
Clarify NSM semantics on function-typed fields in strong mode

### ISSUE THREAD

**Issue by vsmenon** [dart-lang/sdk#30140]
The following exposes a tricky case for supporting old NSM semantics in strong mode:

```dart
class Mock {
  noSuchMethod(Invocation i) {
    return () => 32;
  }
}

class Foo {
  int Function() f;
  String g() => "hello";
}

class Bar extends Mock implements Foo {
}

void baz(Foo foo) {
  print(foo.f);
  print(foo.f());
  print(foo.g());
}

void main() {
  baz(new Bar());
}
```

DDC prints (and throws on the last print):
```
Closure: () -> int from: () => 32
32

/Users/vsm/dart/sdk/pkg/dev_compiler/lib/js/common/dart_sdk.js:9895
          throw e;
          ^
Type '() -> int' is not a subtype of type 'String'
```

The VM in checked mode (dart2js is similar):
```
Closure: () => dynamic
Closure: () => dynamic
Closure: () => dynamic
```

I'd like to propose going forward with the current DDC behavior.  It is much easier to implement in an AOT setting without regressing regular calling conventions.

DDC is able to generate the following for `baz`:

```
mockfield.baz = function(foo) {
  core.print(foo.f);
  core.print(foo.f());
  core.print(foo.g());
};
```

by statically generating trampolines in `Bar` that forward to NSM and type check the result:
```
mockfield.Bar = class Bar extends mockfield.Mock {
  g(...args) {
    return core.String._check(this.noSuchMethod(new dart.InvocationImpl.new('g', args, {isMethod: true})));
  }
  get f() {
    return VoidToint()._check(this.noSuchMethod(new dart.InvocationImpl.new('f', [], {isGetter: true})));
  }
  set f(args) {
    return dart.void._check(this.noSuchMethod(new dart.InvocationImpl.new('f', [args], {isSetter: true})));
  }
};
```

**Comment by leafpetersen** [comment]
The DDC behavior is the intended strong mode behavior.  I would propose that we spec it as the Dart 2.0 behavior.

@lrhn @eernstg @munificent @floitschG

**Comment by munificent** [comment]
I'm fine with the DDC behavior. Over the long term, I think this highlights that "no such *method*" isn't the right primitive for Dart and that "no such *property*" is a better fit.

**Comment by lrhn** [comment]
I'm fine with the DDC behavior too, at least something close to what I think it's doing.
We'll have to nail down the behavior for any call and any declarations, and decide on what happens with invalid invocations. Will `noSuchMethod` get called if you call `foo.g(42)` above? What if you implement two interfaces with different signatures for the same member? Etc.

All in all, I wouldn't mind if the rule is: `noSuchMethod` only get called for interface members with valid arguments (which will likely include type-checking). In all other cases, it'll just throw directly, you don't get to implement something that isn't part of your class' actual signature.
In short, it must be implementable by adding correctly-typed implementations of missing members which forward to `noSuchMethod` when called. No other invocations of `noSuchMethod` is possible.

That means that the following will actually hit `noSuchMethod` in Dart 2:
```dart
class C {
  foo(int x) { print("Yes!"); }
  noSuchMethod(i) { print("No!"); }
}
abstract class D { foo(int x, [int y]); }
class E extends C implements D {}
main() {
  new E().foo();  // prints No!
}
```
(there are likely other cases that we'll have to account for).

**Comment by eernstg** [comment]
The semantics would be as follows: For a class C declaring or inheriting a `noSuchMethod` implementation distinct from the one declared in `Object`, a forwarding member is generated for each member which is declared but not implemented; for a getter/setter/method the forwarding member is a getter/setter/method respectively, and the signature of the forwarder is the signature of the declared but unimplemented member. (There may be conflicts, but that's discussed below.)

With that, the behavior of statically checked member invocations just falls out: Every statically checked member invocation will match a forwarder, both in the sense that it is a getter if a getter is expected, etc., and for methods that the given argument list will be supported by that forwarder method. We don't even have to teach developers to handle tear-offs separately, because a tear-off of a method (statically) will be an actual tear-off of a generated forwarder.

The basic rule could be that `noSuchMethod` may be invoked from these generated forwarding members (no magic here: developers could just as well have written those forwarders themselves, and they just call `noSuchMethod` explicitly), and `noSuchMethod` is _never_ invoked in any other situation.

It makes sense to have this semantics for statically checked invocations.

However, for dynamic invocations it falls short. In particular, it does not support an open-ended set of (emulated) members, and it does not support abstracting over different shapes of argument lists. Actually, the most important problem may be that the resulting semantics of dynamic invocations will be complex and conceptually ill-motivated.

Ideally, the semantics for dynamic invocations should amount to checking whether a given receiver contains a member that admits the given invocation, and invoking `noSuchMethod` when that is not the case (both when the requested member is and when it isn't declared in the type of the receiver, and when the name is there but the argument list does not fit).

That would give dynamic invocations a semantics which preserves a larger portion of the expressive power normally associated with `noSuchMethod`. Crucially, the semantics for statically checked invocations falls out as a special case of the dynamic behavior, so we have a consistent model. It just happens to be a correct simplification & optimization to use forwarders in the statically checked case, because we know positively at compile-time that the member exists and the argument list fits.

Finally, one thing that the forwarder based approach does not handle is the ability to support types that conflict. E.g., forwarders will not enable using `noSuchMethod` to implement types that require both a `foo([int, int])` and a `foo({String x})`, and even with a potential extension of Dart to enable both kinds of optional arguments in the same signature we might have other conflicts (e.g., we may need to have both `int get foo` and `int foo()`). Such conflicts would in fact be supported by a dynamic semantics as proposed.

For static invocations, we would actually be able to detect that a given type offers a conflicting set of declarations for a given member name (if our type system even allows us to express and use such types, including the ability to even declare a `class Mock implements Foo, Bar ..` where `Foo` and `Bar` have a conflict). If so, when we know that a conflicting invocation is requested, we may choose to check that the invocation matches one of the declared signatures and then we may perform the invocation dynamically (maybe we can even prove that this could never be implemented as anything other than `noSuchMethod`, and call that directly).

In any case, the main point is that it would be both natural and helpful to give dynamic invocations a more complete semantics, and I do not think it will cause any performance degradation for the static cases.

**Comment by vsmenon** [comment]
> We'll have to nail down the behavior for any call and any declarations, and decide on what happens with invalid invocations. Will noSuchMethod get called if you call foo.g(42) above?

If the static type of `foo` is `Foo`, then it'd be a static error.  If `dynamic`, a `noSuchMethod` (but that would go through a runtime helper, not the trampoline).

FWIW, when we extend and implement methods with slightly different signatures from multiple types, DDC often has to generate static trampolines - e.g., to inject the right covariance checks.  E.g., this throws a runtime error in `C.foo` - a trampoline with a runtime check before forwarding to `super.foo`.

```
abstract class A<T> {
  void foo(T x);
}

class B {
  void foo(String x) {
    print('hello ' + x);
  }
}

class C extends B implements A<String> {
}

void main() {
  A<Object> a = new C();
  a.foo(42);
}
```

That said, @lrhn 's example is a static error today:

```
[error] Base class introduces an invalid override. The type of 'C.foo' ('(int) → dynamic') isn't a subtype of 'D.foo' ('(int, [int]) → dynamic').
```

**Comment by vsmenon** [comment]
@eernstg - in the original example, if we change `foo` in `baz` to be `dynamic`, I presume the semantics would be the same as the static case?

**Comment by eernstg** [comment]
> if we change foo in baz to be dynamic, I presume the semantics would be the same as the static case?

The first thing to note is that `Bar` will have generated forwarding members corresponding to its declared type. Assuming an actual argument of type `Bar`,

- the invocation `foo.f` would be a dynamic invocation that finds the generated forwarding getter corresponding to `Foo.f` and yields the closure of `() => 32` which successfully passes the return type check (if we make sure that `=>` function literals have a reified type which matches their static type, which is unclear in the current spec);

- `foo.f()` would dynamically fail to find a method `f`, and it would then attempt to invoke a getter `f`, which leads to `noSuchMethod` which yields the same closure, which is called, yielding 32;

- `foo.g()` would dynamically invoke the forwarding method `g` and fail to satisfy the `String` constraint on the value returned from `noSuchMethod`.

We could consider avoiding the invocations of forwarding methods in the dynamic case, but I think that would make the semantics more complex (e.g., because we wouldn't get the type checks on the returned results that static checks would incur). It's a reasonably understandable semantics that `noSuchMethod` will give you these forwarders, and then we just perform dynamic invocations in a manner that includes them.

**Comment by eernstg** [comment]
Lasse's example illustrates one more thing we need to take into account:
```dart
class C {
  foo(int x) { print("Yes!"); }
  noSuchMethod(i) { print("No!"); }
}
abstract class D { foo(int x, [int y]); }
class E extends C implements D {}
main() {
  new E().foo();  // prints No!
}
```
The semantics should ensure that an invocation of a method which is actually implemented by the receiver must go to that implementation, not to `noSuchMethod`.

This means that if we generate a forwarder for `foo` in `E` such that the required interface is actually implemented, and if that forwarder overrides an existing implementation, then invocations which may call that implementation must call that implementation:
```dart
class E extends C implements D {
  // Generated forwarder.
  foo(int x, [int y]) {
    if (/* exactly one argument was passed */) {
      return super.foo(x);
    } else {
      return noSuchMethod(new Invocation.inventingAnAccessibleConstructor(
          memberName: #foo, positionalArguments: [x, y], isMethod: true));
    }
  }
}
```

**Comment by eernstg** [comment]
An 'intent to implement' issue has now been created for the language design which was specified in the [informal specification](https://github.com/dart-lang/sdk/blob/master/docs/language/informal/nosuchmethod-forwarding.md) of the feature arising from the discussions that took place in this issue as well as elsewhere. I've referred to this issue as background material in #31426. Closing.
